File: Advent19.java
import java.util.*;
import java.util.stream.*;

public class Advent19 extends Advent {

  int pc = 0;
  int pcReg;
  int[] regs;
  List<String> instructions;

  public Advent19() {
    super(19);
  }

  @Override
  protected void parseInput() {
    pc = 0;
    regs = new int[6];
    instructions = new ArrayList<>();
    for (String line : input) {
      if (line.startsWith("#")) {
        pcReg = Integer.parseInt(line.split(" ")[1]);
      } else {
        instructions.add(line);
      }
    }
  }

  @Override
  protected String part1() {
    runProgram();
    return "" + regs[0];
  }

  private void runProgram() {
    while(pc < instructions.size()) {
      regs[pcReg] = pc;
      eval(instructions.get(pc));
      pc = regs[pcReg];
      pc++;
    }
  }

  private void eval(String instruction) {
    String[] groups = instruction.split(" ");
    String op = groups[0];
    int a = Integer.parseInt(groups[1]);
    int b = Integer.parseInt(groups[2]);
    int c = Integer.parseInt(groups[3]);
    eval(op, a, b, c);
  }

  private void eval(String op, int a, int b, int c) {
    switch(op) {
      case "addr":
        addr(a, b, c); break;
      case "addi":
        addi(a, b, c); break;
      case "mulr":
        mulr(a, b, c); break;
      case "muli":
        muli(a, b, c); break;
      case "banr":
        banr(a, b, c); break;
      case "bani":
        bani(a, b, c); break;
      case "borr":
        borr(a, b, c); break;
      case "bori":
        bori(a, b, c); break;
      case "setr":
        setr(a, b, c); break;
      case "seti":
        seti(a, b, c); break;
      case "gtir":
        gtir(a, b, c); break;
      case "gtri":
        gtri(a, b, c); break;
      case "gtrr":
        gtrr(a, b, c); break;
      case "eqir":
        eqir(a, b, c); break;
      case "eqri":
        eqri(a, b, c); break;
      case "eqrr":
        eqrr(a, b, c); break;
      default:
        throw new UnsupportedOperationException();
    }
  }

  private void addr(int a, int b, int c) {
    regs[c] = regs[a] + regs[b];
  }

  private void addi(int a, int b, int c) {
    regs[c] = regs[a] + b;
  }

  private void mulr(int a, int b, int c) {
    regs[c] = regs[a]*regs[b];
  }

  private void muli(int a, int b, int c) {
    regs[c] = regs[a]*b;
  }

  private void banr(int a, int b, int c) {
    regs[c] = regs[a] & regs[b];
  }

  private void bani(int a, int b, int c) {
    regs[c] = regs[a] & b;
  }

  private void borr(int a, int b, int c) {
    regs[c] = regs[a] | regs[b];
  }

  private void bori(int a, int b, int c) {
    regs[c] = regs[a] | b;
  }

  private void setr(int a, int b, int c) {
    regs[c] = regs[a];
  }

  private void seti(int a, int b, int c) {
    regs[c] = a;
  }

  private void gtir(int a, int b, int c) {
    regs[c] = a > regs[b] ? 1 : 0;
  }

  private void gtri(int a, int b, int c) {
    regs[c] = regs[a] > b ? 1 : 0;
  }

  private void gtrr(int a, int b, int c) {
    regs[c] = regs[a] > regs[b] ? 1 : 0;
  }

  private void eqir(int a, int b, int c) {
    regs[c] = a == regs[b] ? 1 : 0;
  }

  private void eqri(int a, int b, int c) {
    regs[c] = regs[a] == b ? 1 : 0;
  }

  private void eqrr(int a, int b, int c) {
    regs[c] = regs[a] == regs[b] ? 1 : 0;
  }

  @Override
  protected String part2() {
    /*
    This day was about reverse engineering. It sets the final register
    to a large number, and asks us to sum its factors. For my input, the
    large number was 10551329.
    */
    int result = 0;
    int n = 10551329;
    for (int i = 1; i <= n; i++) {
      if (n % i == 0) {
        result += i;
      }
    }
    return "" + result;
  }
}

File: 2018/python/advent_19.py
from common.advent_lib import *
from collections import Counter, defaultdict, deque
import collections as coll
import functools as func
import itertools as it
import operator as oper
import re
import more_itertools as it2
import numpy as np
import sympy as sym

lines = read_lines(19)

def addr(a,b,c,regs):
    new_regs=regs.copy()
    new_regs[c]=new_regs[a]+new_regs[b]
    return new_regs

def addi(a,b,c,regs):
    new_regs=regs.copy()
    new_regs[c]=new_regs[a]+b
    return new_regs

def mulr(a,b,c,regs):
    new_regs=regs.copy()
    new_regs[c]=new_regs[a]*new_regs[b]
    return new_regs

def muli(a,b,c,regs):
    new_regs=regs.copy()
    new_regs[c]=new_regs[a]*b
    return new_regs

def banr(a,b,c,regs):
    new_regs=regs.copy()
    new_regs[c]=new_regs[a]&new_regs[b]
    return new_regs

def bani(a,b,c,regs):
    new_regs=regs.copy()
    new_regs[c]=new_regs[a]&b
    return new_regs

def borr(a,b,c,regs):
    new_regs=regs.copy()
    new_regs[c]=new_regs[a]|new_regs[b]
    return new_regs

def bori(a,b,c,regs):
    new_regs=regs.copy()
    new_regs[c]=new_regs[a]|b
    return new_regs

def setr(a,_,c,regs):
    new_regs=regs.copy()
    new_regs[c]=new_regs[a]
    return new_regs

def seti(a,_,c,regs):
    new_regs=regs.copy()
    new_regs[c]=a
    return new_regs

def gtir(a,b,c,regs):
    new_regs=regs.copy()
    new_regs[c]=int(a>new_regs[b])
    return new_regs

def gtri(a,b,c,regs):
    new_regs=regs.copy()
    new_regs[c]=int(new_regs[a]>b)
    return new_regs

def gtrr(a,b,c,regs):
    new_regs=regs.copy()
    new_regs[c]=int(new_regs[a]>new_regs[b])
    return new_regs

def eqir(a,b,c,regs):
    new_regs=regs.copy()
    new_regs[c]=int(a==new_regs[b])
    return new_regs

def eqri(a,b,c,regs):
    new_regs=regs.copy()
    new_regs[c]=int(new_regs[a]==b)
    return new_regs

def eqrr(a,b,c,regs):
    new_regs=regs.copy()
    new_regs[c]=int(new_regs[a]==new_regs[b])
    return new_regs

program_table = {f.__name__:f for f in [addr, addi, mulr, muli, banr, bani, borr, bori, setr, seti, gtir, gtri, gtrr, eqir, eqri, eqrr]}
IP=int(lines[0].split()[-1])
regs=[0]*6

def evaluate(program, regs):
    op,*args=program[regs[IP]].split()
    a,b,c=map(int,args)
    return program_table[op](a,b,c,regs)

while regs[IP]<len(lines):
    new_regs=evaluate(lines[1:],regs)
    new_regs[IP]+=1
    regs=new_regs

part1 = regs[0]
print("Part 1:", part1)

regs=[0]*6
regs[0]=1

def pretty_regs(regs):
    expanded_regs=[f"{reg:8d}" for reg in regs]
    return f"[{', '.join(expanded_regs)}]"

def pretty_op(instr):
    op,*args=instr.split()
    a,b,c=map(int,args)
    return f"{op} {a:2d} {b:2d} {c:2d}"

breakpoints=["regs[5]==10551329"]
def run_program(program, regs,ip,breakpoints=[]):
    breakpoint_triggered=False
    while regs[ip]<len(program):
        print(f"ip={regs[ip]:2d} {pretty_regs(regs)} {pretty_op(program[regs[ip]])}")
        for bp in breakpoints:
            if eval(bp):
                breakpoint_triggered=True
        if breakpoint_triggered:
            _input=input("Command: ")
            while _input:
                if _input=="clear":
                    breakpoint_triggered=False
                else:
                    if _input.split()[0] not in program_table.keys():
                        _input=input(f"Not a valid command ({_input}):")
                    else:
                        op,*args=_input.split()
                        a,b,c=map(int,args)
                        regs = program_table[op](a,b,c,regs)

        new_regs=evaluate(program, regs)
        # input()
        new_regs[ip]+=1
        regs=new_regs

# The assembly code does an inefficient search of divisors of N
# if register 0 bit is set, N = 10551329, else N = 929
# The reverse engineered algorithm is as follows:
#
# N = 10551329 if $0 else 929
# sum_of_divisors = 0
# for i in range(1,N):
#   for j in range(1,N):
#       if i*j=N:
#           sum_of_divisors+=i
# $0 = sum_of_divisors

part2 = sum(sym.divisors(10551329))
print("Part 2:", part2)
